mapproxy/source/mapnik.py

```diff
@@ -93,24 +96,42 @@ def render(self, query):
         else:
             return self.render_mapfile(mapfile, query)
 
-    def map_obj(self, mapfile):
-        if not self._cache_map_obj:
-            m = mapnik.Map(0, 0)
-            mapnik.load_map(m, str(mapfile))
-            return m
+    def _create_map_obj(self, mapfile):
+        m = mapnik.Map(0, 0)
+        mapnik.load_map(m, str(mapfile))
+        return m
 
+    def map_obj(self, mapfile):
         # cache loaded map objects
-        # only works when a single proc/thread accesses this object
+        # only works when a single proc/thread accesses the map
         # (forking the render process doesn't work because of open database
         #  file handles that gets passed to the child)
-        if mapfile not in self._map_objs:
-            with self._map_objs_lock:
-                if mapfile not in self._map_objs:
-                    m = mapnik.Map(0, 0)
-                    mapnik.load_map(m, str(mapfile))
-                    self._map_objs[mapfile] = m
-
-        return self._map_objs[mapfile]
+        # segment the cache by process and thread to avoid interference
+        if self._cache_map_obj:
+            cachekey = None # renderd guarantees that there are no concurrency issues
+        else:
+            thread_id = threading.current_thread().ident
+            process_id = multiprocessing.current_process()._identity
+            cachekey = (process_id, thread_id, mapfile)
+        if cachekey not in _map_objs:
```

Comment:
You can reuse map objects across multiple threads. Just take them out of the _map_obj during use and push them back afterwards. Make sure to use locking (threading.Lock) when accessing _map_obj. And also store the PID with the map object and verify that the PID matches the current PID.

Reply:
This sounds like the best tool to use here would be a dict of queues identified by the PID, with the cleanup pushing map objects back into the queue. Does that sound about right?

Reply:
I now implemented queues for unused mapnik objects. I’ll still need help in adding a clean test for this.

I found test/unit/test_async.py, but I am not sure how to use it correctly to test the specific mapnik instance with threading.

Reply:
I now added an additional check to ensure that the PID always matches the PID of the process when retrieving a mapfile.

I see test failures for that commit, but the same test failures [currently occur on master](https://travis-ci.org/mapproxy/mapproxy/builds/559887389):
```
mapproxy/test/unit/test_cache_riak.py::TestRiakCacheHTTP::test_double_remove FAILED [ 48%]
```